### PR TITLE
tarsum: init

### PR DESCRIPTION
**Pull request by zowoq**
###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Relase notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by zowoq** [comment]
@roberth Should `tarsum` be moved to a package so it is evaled/built on ofborg/hydra? It seems a bit odd that we're treating it like a script instead of a binary?

**Comment by roberth** [comment]
It is built indirectly by hydra, but not on ofborg by default. I'm not aware of any use cases for tarsum other than dockerTools, but my imagination is limited and making it more like a regular package is good for maintenance.

@GrahamcOfBorg build dockerTools.tarsum

^ first and last time I write this, I suppose :)

File: pkgs/build-support/docker/default.nix [hunk 1]
```diff
@@ -125,6 +125,7 @@ rec {
   # And we cannot untar it, because then we cannot preserve permissions etc.
   tarsum = runCommand "tarsum" {
     nativeBuildInputs = [ go ];
+    disallowedReferences = [ go ];
```
Comment (roberth) [inline pkgs/build-support/docker/default.nix:128]: Instance of https://github.com/NixOS/nixpkgs/issues/132637 (no action required here now)

File: pkgs/build-support/docker/tarsum.nix [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+  buildPhase = ''
+    runHook preBuild
+    mkdir tarsum
+    cd tarsum
+    cp ${./tarsum.go} tarsum.go
+    export GOPATH=$(pwd)
+    export GOCACHE="$TMPDIR/go-cache"
+    mkdir -p src/github.com/docker/docker/pkg
+    ln -sT ${docker.moby-src}/pkg/tarsum src/github.com/docker/docker/pkg/tarsum
```
Comment (zowoq) [inline pkgs/build-support/docker/tarsum.nix:23]: Symlinking the source isn't really compatible with `buildGo{Module,Package}` so I'll leave it as it is for now.

We'll likely need to change this to use go modules and version it when `go` deprecates `GO111MODULE = off` but I don't see much benefit in doing that until it's necessary.

File: pkgs/build-support/docker/tarsum.nix [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+    runHook postBuild
+  '';
+
+  installPhase = ''
+    runHook preInstall
+    mkdir -p $out/bin
+    cp tarsum $out/bin/
+    runHook postInstall
+  '';
+
```
Comment (roberth) [inline pkgs/build-support/docker/tarsum.nix:34]:
```suggestion

  passthru = {
    tests = {
      dockerTools = nixosTests.docker-tools;
    };
  };

```

**Review by roberth — COMMENTED** [review]
Looks decent, but it's still marked as draft.

**Review by roberth — APPROVED** [review]

**Comment by roberth** [comment]
Thanks!

**Pull request merged**